jbmc/src/java_bytecode/README.md

```diff
@@ -736,3 +736,151 @@ See also \ref string-solver-interface.
 \section java-bytecode-conversion-example-section A worked example of converting java bytecode to codet
 
 To be documented.
+
+\section java-trace Java Trace Assumptions
+
+For high-level documentation about goto traces, see \ref goto-trace-structure.
+
+There are three types of assignments to construct all types of Java objects:
+- variable assignments (\ref symbol_exprt)
+- member assignments (\ref member_exprt)
+- index assignments (\ref index_exprt)
+
+We use the convention left-hand side (LHS) = right-hand side (RHS) to represent
+assignments in this document.
+
+The expressions on each side vary depending on the type, whether its a
+member assignment, whether its an assignment by reference, etc. Some examples:
+- A global variable assignment will have a LHS \ref symbol_exprt containing the
+type information and an identifier that does not include a function
+identifier, e.g. `java::SomeClass.someGlobalField`, and a RHS expression
+containing the value, e.g. a \ref constant_exprt, a \ref struct_exprt or an
+\ref address_of_exprt.
+See \ref java-trace-variable-assignment "Variable assignments".
+- A local variable assignment will have a LHS \ref symbol_exprt containing the
+type information and an identifier that includes a function identifier, e.g.
+a parameter assignment like `java::SomeClass.<init>:()V::this` which corresponds
+to a symbol with `is_static_lifetime` set to true.
+See \ref java-trace-variable-assignment "Variable assignments".
+- A member assignment will have a LHS \ref member_exprt containing the type
+information for the member, the component name (member name), and an operand for
+the containing class which contains an identifier for the containing class, a
+RHS expression containing the value. Inherited members have a nested member
+structure. Members of members are assigned using a series of single member
+assignments. See \ref java-trace-member-assignment "Member assignments".
+- An array assignment consists of first assigning to a LHS \ref symbol_exprt
+a RHS of type `java::array`, then if the array is non-empty it is followed by a
+series of \ref index_exprt assignments to assign elements to the array of the
+required type. A notable exception to this assignment structure is when arrays
+are created non-deterministically for primitive types - these are created by
+assigning an entire array to the `java::array` symbol instead of being built
+using index assignments.
```

Comment:
🤔 I wonder if just putting an example goto trace printout could be better than writing out descriptions like this. I'm lost after reading this. But that just might be me...

Also, as far as I can tell you didn't write this, so not a big problem.

Reply:
Actually, I did write it in another PR, but I just changed the location in this PR. I reckon that it is more interesting to know that arrays constructed by making a struct then assigning indices than it is to look at a print out here. That example belongs in the array section, in my opinion @thk123 may want to update this text to reflect his documentation efforts though.